File: node/src/rpc/acceptance.test.ts
// reflect-metadata must load before @peculiar/x509: it pulls in tsyringe, which
// hard-fails without the polyfill depending on module load order.
import "reflect-metadata";
import { beforeAll, describe, expect, it } from "bun:test";
import * as x509 from "@peculiar/x509";
import type {
	PolicyEvaluation,
	PolicyRule,
} from "../pb/servicebridge/v1/registry";
import {
	checkAcceptance,
	classifyPeer,
	evaluatePeerAcceptance,
	getPeerCertFromCall,
	parsePeerSpiffeUri,
	RUNTIME_PEER_COMMON_NAME,
} from "./acceptance";

const callerA = "aaaaaaaa-0000-0000-0000-000000000001";
const callerB = "bbbbbbbb-0000-0000-0000-000000000002";

// ── real certificate fixtures ────────────────────────────────────────────────
// The acceptance decision reads the DER, so the tests issue real certificates
// rather than hand-shaped objects. Anything less would not exercise the SAN
// parsing that the decision actually depends on.

interface CertFixture {
	raw: Uint8Array;
	subjectaltname?: string;
	subject: { CN?: string };
}

async function issueCert(
	commonName: string,
	uriSans: string[],
): Promise<CertFixture> {
	const alg = { name: "ECDSA", namedCurve: "P-256", hash: "SHA-256" };
	const keys = (await crypto.subtle.generateKey(alg, false, [
		"sign",
		"verify",
	])) as CryptoKeyPair;
	const extensions =
		uriSans.length > 0
			? [
					new x509.SubjectAlternativeNameExtension(
						uriSans.map((value) => ({ type: "url" as const, value })),
					),
				]
			: [];
	const cert = await x509.X509CertificateGenerator.createSelfSigned({
		name: `CN=${commonName}`,
		keys,
		extensions,
		signingAlgorithm: alg,
	});
	return {
		raw: new Uint8Array(cert.rawData),
		subjectaltname:
			uriSans.length > 0
				? uriSans.map((u) => `URI:${u}`).join(", ")
				: undefined,
		subject: { CN: commonName },
	};
}

function spiffeUri(serviceId: string, instanceId = "i-1"): string {
	return `spiffe://service-bridge/service/${serviceId}/instance/${instanceId}`;
}

let sdkLeafA: CertFixture;
let sdkLeafB: CertFixture;
let runtimeCert: CertFixture;
let foreignUriCert: CertFixture;
let noSanStrangerCert: CertFixture;

beforeAll(async () => {
	sdkLeafA = await issueCert("servicebridge-leaf", [spiffeUri(callerA)]);
	sdkLeafB = await issueCert("servicebridge-leaf", [spiffeUri(callerB)]);
	runtimeCert = await issueCert(RUNTIME_PEER_COMMON_NAME, []);
	foreignUriCert = await issueCert("servicebridge-leaf", [
		"https://attacker.example/service/x",
	]);
	noSanStrangerCert = await issueCert("some-other-service", []);
});

// makeCall builds the real grpc-js server-call shape:
// ServerUnaryCall.call → BaseServerInterceptingCall.stream → session.socket.
function makeCall(cert: CertFixture): object {
	return {
		call: {
			stream: {
				session: {
					socket: { getPeerCertificate: () => cert },
				},
			},
		},
	};
}

// makeAuthContextCall builds the public grpc-js accessor shape.
function makeAuthContextCall(cert: CertFixture): object {
	return {
		getAuthContext: () => ({
			transportSecurityType: "ssl",
			sslPeerCertificate: cert,
		}),
	};
}

// ── parsePeerSpiffeUri ───────────────────────────────────────────────────────

describe("parsePeerSpiffeUri", () => {
	it("parses valid SPIFFE URI", () => {
		const result = parsePeerSpiffeUri(spiffeUri(callerA, callerB));
		expect(result).toEqual({ serviceId: callerA, instanceId: callerB });
	});

	it("returns null for non-SPIFFE URI", () => {
		expect(parsePeerSpiffeUri("https://example.com")).toBeNull();
	});

	it("returns null for SPIFFE URI with wrong structure", () => {
		expect(parsePeerSpiffeUri("spiffe://service-bridge/foo/bar")).toBeNull();
	});

	it("returns null for a SPIFFE URI from another trust domain", () => {
		expect(
			parsePeerSpiffeUri("spiffe://other/service/svc-a/instance/i-1"),
		).toBeNull();
	});

	it("returns null when serviceId or instanceId is empty", () => {
		expect(
			parsePeerSpiffeUri("spiffe://service-bridge/service//instance/i-1"),
		).toBeNull();
		expect(
			parsePeerSpiffeUri("spiffe://service-bridge/service/svc-a/instance/"),
		).toBeNull();
	});
});

// ── classifyPeer ─────────────────────────────────────────────────────────────

describe("classifyPeer", () => {
	it("identifies an SDK peer from the DER SPIFFE URI SAN", () => {
		expect(classifyPeer(sdkLeafA)).toEqual({
			kind: "service",
			serviceId: callerA,
		});
	});

	it("identifies the SDK peer even when Node leaves subjectaltname empty", () => {
		// Node does this for certs whose SAN carries only URI entries — the DER
		// is what makes the identity recoverable.
		expect(
			classifyPeer({
				raw: sdkLeafA.raw,
				subject: { CN: "servicebridge-leaf" },
			}),
		).toEqual({ kind: "service", serviceId: callerA });
	});

	it("identifies the runtime by CN when the cert has no URI SAN", () => {
		expect(classifyPeer(runtimeCert)).toEqual({ kind: "runtime" });
	});

	it("rejects a cert whose URI SAN is not a ServiceBridge SPIFFE identity", () => {
		const peer = classifyPeer(foreignUriCert);
		expect(peer.kind).toBe("unknown");
		expect(peer.kind === "unknown" && peer.reason).toContain("URI SAN");
	});

	it("rejects a SAN-less cert that is not the runtime", () => {
		const peer = classifyPeer(noSanStrangerCert);
		expect(peer.kind).toBe("unknown");
	});

	it("rejects an unparseable DER instead of assuming the runtime", () => {
		const peer = classifyPeer({
			raw: new Uint8Array([1, 2, 3, 4]),
			subject: { CN: RUNTIME_PEER_COMMON_NAME },
		});
		expect(peer.kind).toBe("unknown");
		expect(peer.kind === "unknown" && peer.reason).toContain("unparseable");
	});

	it("rejects an empty cert object", () => {
		expect(classifyPeer({}).kind).toBe("unknown");
	});

	it("falls back to subjectaltname when no DER is present", () => {
		expect(
			classifyPeer({ subjectaltname: `URI:${spiffeUri(callerA)}` }),
		).toEqual({ kind: "service", serviceId: callerA });
	});

	it("reads the runtime CN from subject when no DER is present", () => {
		expect(classifyPeer({ subject: { CN: RUNTIME_PEER_COMMON_NAME } })).toEqual(
			{ kind: "runtime" },
		);
	});

	it("picks the SPIFFE URI out of a multi-entry SAN", () => {
		expect(
			classifyPeer({
				subjectaltname: `DNS:host.local, URI:${spiffeUri(callerB)}`,
			}),
		).toEqual({ kind: "service", serviceId: callerB });
	});

	it("does not treat a DNS-only SAN as the runtime when CN says otherwise", () => {
		expect(
			classifyPeer({
				subjectaltname: "DNS:host.local",
				subject: { CN: "servicebridge-leaf" },
			}).kind,
		).toBe("unknown");
	});
});

// ── checkAcceptance ──────────────────────────────────────────────────────────

function makeAcceptanceRule(
	peerServiceId: string,
	targetName: string,
): PolicyRule {
	return {
		action: "rpc.handle",
		peerServiceId,
		peerServiceName: "",
		targetName,
	};
}

function policyWith(rules: PolicyRule[]): PolicyEvaluation {
	return { capabilities: [], egress: [], acceptance: rules, warnings: [] };
}

describe("checkAcceptance", () => {
	it("default-allow when no acceptance rules", () => {
		expect(checkAcceptance(callerA, "charge", policyWith([]))).toBe(true);
	});

	it("allows matching peer + exact method", () => {
		const policy = policyWith([makeAcceptanceRule(callerA, "charge")]);
		expect(checkAcceptance(callerA, "charge", policy)).toBe(true);
	});

	it("allows matching peer + wildcard method", () => {
		const policy = policyWith([makeAcceptanceRule(callerA, "*")]);
		expect(checkAcceptance(callerA, "anything", policy)).toBe(true);
	});

	it("denies unlisted peer", () => {
		const policy = policyWith([makeAcceptanceRule(callerA, "charge")]);
		expect(checkAcceptance(callerB, "charge", policy)).toBe(false);
	});

	it("denies listed peer with wrong method", () => {
		const policy = policyWith([makeAcceptanceRule(callerA, "charge")]);
		expect(checkAcceptance(callerA, "refund", policy)).toBe(false);
	});

	it("allows when peer matches a wildcard-peer rule (empty peerServiceId)", () => {
		const policy = policyWith([makeAcceptanceRule("", "charge")]);
		expect(checkAcceptance(callerB, "charge", policy)).toBe(true);
	});

	it("ignores acceptance rules for other actions", () => {
		const policy = policyWith([
			{
				action: "event.subscribe",
				peerServiceId: callerA,
				peerServiceName: "",
				targetName: "charge",
			},
		]);
		expect(checkAcceptance(callerB, "charge", policy)).toBe(true);
	});
});

// ── getPeerCertFromCall ──────────────────────────────────────────────────────

describe("getPeerCertFromCall", () => {
	it("reads the cert through the public getAuthContext accessor", () => {
		const cert = getPeerCertFromCall(makeAuthContextCall(sdkLeafA));
		expect(cert?.raw).toEqual(sdkLeafA.raw);
	});

	it("walks call.stream.session.socket when getAuthContext is absent", () => {
		const cert = getPeerCertFromCall(makeCall(sdkLeafA));
		expect(cert?.raw).toEqual(sdkLeafA.raw);
	});

	it("walks a top-level stream.session.socket", () => {
		const cert = getPeerCertFromCall({
			stream: { session: { socket: { getPeerCertificate: () => sdkLeafA } } },
		});
		expect(cert?.raw).toEqual(sdkLeafA.raw);
	});

	it("falls through to the socket walk when getAuthContext yields no cert", () => {
		// grpc-js returns {} for a non-TLS socket.
		const call = {
			getAuthContext: () => ({}),
			call: {
				stream: {
					session: { socket: { getPeerCertificate: () => sdkLeafA } },
				},
			},
		};
		expect(getPeerCertFromCall(call)?.raw).toEqual(sdkLeafA.raw);
	});

	it("returns null when no accessor is present", () => {
		expect(getPeerCertFromCall({})).toBeNull();
		expect(getPeerCertFromCall({ call: {} })).toBeNull();
		expect(getPeerCertFromCall({ call: { stream: {} } })).toBeNull();
	});

	it("returns null when the session has no socket", () => {
		expect(
			getPeerCertFromCall({ call: { stream: { session: {} } } }),
		).toBeNull();
	});

	it("returns null when getPeerCertificate is not a function", () => {
		expect(
			getPeerCertFromCall({
				call: { stream: { session: { socket: { getPeerCertificate: 42 } } } },
			}),
		).toBeNull();
	});

	it("returns null when getPeerCertificate throws", () => {
		expect(
			getPeerCertFromCall({
				call: {
					stream: {
						session: {
							socket: {
								getPeerCertificate: () => {
									throw new Error("socket closed");
								},
							},
						},
					},
				},
			}),
		).toBeNull();
	});
});

// ── evaluatePeerAcceptance ───────────────────────────────────────────────────

describe("evaluatePeerAcceptance", () => {
	const restrictivePolicy = policyWith([makeAcceptanceRule(callerA, "charge")]);

	it("default-allow when policy is null", () => {
		expect(evaluatePeerAcceptance(null, makeCall(sdkLeafA), "x")).toBeNull();
	});

	it("default-allow when policy has no rpc.handle rules", () => {
		expect(
			evaluatePeerAcceptance(policyWith([]), makeCall(sdkLeafA), "x"),
		).toBeNull();
	});

	it("fail-closed when the peer certificate cannot be extracted", () => {
		const denial = evaluatePeerAcceptance(restrictivePolicy, {}, "charge");
		expect(denial).toContain("could not extract peer certificate");
	});

	it("fail-closed when the cert accessor throws", () => {
		const call = {
			getAuthContext: () => {
				throw new Error("no auth context");
			},
		};
		const denial = evaluatePeerAcceptance(restrictivePolicy, call, "charge");
		expect(denial).toContain("could not extract peer certificate");
	});

	it("fail-closed when the peer identity cannot be established", () => {
		const denial = evaluatePeerAcceptance(
			restrictivePolicy,
			makeCall(noSanStrangerCert),
			"charge",
		);
		expect(denial).toContain("peer identity not established");
	});

	it("fail-closed on an unresolvable URI SAN instead of assuming the runtime", () => {
		const denial = evaluatePeerAcceptance(
			restrictivePolicy,
			makeCall(foreignUriCert),
			"charge",
		);
		expect(denial).toContain("peer identity not established");
	});

	it("fail-closed on an unparseable peer certificate", () => {
		const denial = evaluatePeerAcceptance(
			restrictivePolicy,
			makeCall({
				raw: new Uint8Array([9, 9, 9]),
				subject: { CN: RUNTIME_PEER_COMMON_NAME },
			} as CertFixture),
			"charge",
		);
		expect(denial).toContain("unparseable");
	});

	it("allows the runtime proxy, which enforced gate #3 already", () => {
		expect(
			evaluatePeerAcceptance(
				restrictivePolicy,
				makeCall(runtimeCert),
				"charge",
			),
		).toBeNull();
	});

	it("denies when acceptance rules reject the caller", () => {
		const denial = evaluatePeerAcceptance(
			restrictivePolicy,
			makeCall(sdkLeafB),
			"charge",
		);
		expect(denial).toContain("acceptance denied");
		expect(denial).toContain(callerB);
	});

	it("denies an allowed caller invoking a method it may not call", () => {
		const denial = evaluatePeerAcceptance(
			restrictivePolicy,
			makeCall(sdkLeafA),
			"refund",
		);
		expect(denial).toContain("acceptance denied");
	});

	it("allows when acceptance rules accept the caller", () => {
		expect(
			evaluatePeerAcceptance(restrictivePolicy, makeCall(sdkLeafA), "charge"),
		).toBeNull();
	});

	it("allows through the public getAuthContext accessor too", () => {
		expect(
			evaluatePeerAcceptance(
				restrictivePolicy,
				makeAuthContextCall(sdkLeafA),
				"charge",
			),
		).toBeNull();
	});
});

File: go/internal/rpc/acceptance_test.go
package rpc

import (
	"context"
	"crypto/ecdsa"
	"crypto/elliptic"
	"crypto/rand"
	"crypto/tls"
	"crypto/x509"
	"crypto/x509/pkix"
	"errors"
	"math/big"
	"net/url"
	"testing"
	"time"

	"github.com/service-bridge/sdk/go/internal/connection"
	pb "github.com/service-bridge/sdk/go/internal/pb/servicebridge/v1"
)

// inboundCA mirrors the runtime's CA: it signs the SPIFFE leaves handed to SDK
// instances and the SAN-less leaf the runtime presents when it proxies.
type inboundCA struct {
	cert *x509.Certificate
	der  []byte
	key  *ecdsa.PrivateKey
}

func newInboundCA(t *testing.T) *inboundCA {
	t.Helper()

	key, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
	if err != nil {
		t.Fatalf("generate CA key: %v", err)
	}
	tmpl := &x509.Certificate{
		SerialNumber:          certSerial(t),
		Subject:               pkix.Name{CommonName: "servicebridge-ca"},
		NotBefore:             time.Now().Add(-time.Minute),
		NotAfter:              time.Now().Add(time.Hour),
		KeyUsage:              x509.KeyUsageCertSign | x509.KeyUsageDigitalSignature,
		BasicConstraintsValid: true,
		IsCA:                  true,
	}
	der, err := x509.CreateCertificate(rand.Reader, tmpl, tmpl, &key.PublicKey, key)
	if err != nil {
		t.Fatalf("create CA cert: %v", err)
	}
	cert, err := x509.ParseCertificate(der)
	if err != nil {
		t.Fatalf("parse CA cert: %v", err)
	}
	return &inboundCA{cert: cert, der: der, key: key}
}

func (ca *inboundCA) issue(t *testing.T, tmpl *x509.Certificate) tls.Certificate {
	t.Helper()

	key, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
	if err != nil {
		t.Fatalf("generate leaf key: %v", err)
	}
	tmpl.SerialNumber = certSerial(t)
	tmpl.NotBefore = time.Now().Add(-time.Minute)
	tmpl.NotAfter = time.Now().Add(time.Hour)
	tmpl.KeyUsage = x509.KeyUsageDigitalSignature
	tmpl.ExtKeyUsage = []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth}

	der, err := x509.CreateCertificate(rand.Reader, tmpl, ca.cert, &key.PublicKey, ca.key)
	if err != nil {
		t.Fatalf("create leaf: %v", err)
	}
	leaf, err := x509.ParseCertificate(der)
	if err != nil {
		t.Fatalf("parse leaf: %v", err)
	}
	return tls.Certificate{
		Certificate: [][]byte{der, ca.der},
		PrivateKey:  key,
		Leaf:        leaf,
	}
}

// serviceLeaf issues the shape every SDK instance gets: a SPIFFE URI SAN.
func (ca *inboundCA) serviceLeaf(t *testing.T, serviceID, instanceID string) tls.Certificate {
	t.Helper()

	uri, err := url.Parse(connection.FormatSPIFFE(connection.Identity{ServiceID: serviceID, InstanceID: instanceID}))
	if err != nil {
		t.Fatalf("parse SPIFFE URI: %v", err)
	}
	return ca.issue(t, &x509.Certificate{
		Subject: pkix.Name{CommonName: "servicebridge-leaf"},
		URIs:    []*url.URL{uri},
	})
}

// runtimeLeaf issues the shape the runtime presents when it proxies: no SAN at
// all, identified only by its CN.
func (ca *inboundCA) runtimeLeaf(t *testing.T) tls.Certificate {
	t.Helper()
	return ca.issue(t, &x509.Certificate{Subject: pkix.Name{CommonName: RuntimeCommonName}})
}

func (ca *inboundCA) sanlessLeaf(t *testing.T, commonName string) tls.Certificate {
	t.Helper()
	return ca.issue(t, &x509.Certificate{Subject: pkix.Name{CommonName: commonName}})
}

func (ca *inboundCA) foreignURILeaf(t *testing.T, raw string) tls.Certificate {
	t.Helper()

	uri, err := url.Parse(raw)
	if err != nil {
		t.Fatalf("parse URI %q: %v", raw, err)
	}
	return ca.issue(t, &x509.Certificate{
		Subject: pkix.Name{CommonName: "servicebridge-leaf"},
		URIs:    []*url.URL{uri},
	})
}

func (ca *inboundCA) credentials(t *testing.T, leaf tls.Certificate, serviceID, instanceID string) connection.Credentials {
	t.Helper()
	return connection.Credentials{
		Addr: "127.0.0.1:0",
		Lease: connection.Lease{
			Identity:   connection.Identity{ServiceID: serviceID, InstanceID: instanceID},
			CertDER:    leaf.Certificate[0],
			CAChainDER: ca.der,
			TLSCert:    leaf,
		},
		TLS: connection.MutualTLSConfig(ca.cert, leaf),
	}
}

func certSerial(t *testing.T) *big.Int {
	t.Helper()
	n, err := rand.Int(rand.Reader, new(big.Int).Lsh(big.NewInt(1), 128))
	if err != nil {
		t.Fatalf("certSerial: %v", err)
	}
	return n
}

func TestIdentifyPeer(t *testing.T) {
	t.Parallel()

	ca := newInboundCA(t)

	tests := []struct {
		name     string
		cert     *x509.Certificate
		wantKind PeerKind
		wantID   string
		wantErr  bool
	}{
		{
			name:     "SPIFFE leaf is a service",
			cert:     ca.serviceLeaf(t, "svc-a", "inst-1").Leaf,
			wantKind: PeerService,
			wantID:   "svc-a",
		},
		{
			name:     "SAN-less runtime CN is the proxy",
			cert:     ca.runtimeLeaf(t).Leaf,
			wantKind: PeerRuntime,
		},
		{
			name:    "URI SAN that is not SPIFFE is refused",
			cert:    ca.foreignURILeaf(t, "https://example.test/whoami").Leaf,
			wantErr: true,
		},
		{
			name:    "SPIFFE URI in another trust domain is refused",
			cert:    ca.foreignURILeaf(t, "spiffe://elsewhere/service/svc-a/instance/inst-1").Leaf,
			wantErr: true,
		},
		{
			name:    "SAN-less leaf with a foreign CN is refused",
			cert:    ca.sanlessLeaf(t, "servicebridge-leaf").Leaf,
			wantErr: true,
		},
		{
			name:    "no certificate is refused",
			cert:    nil,
			wantErr: true,
		},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			t.Parallel()

			got, err := IdentifyPeer(tc.cert)
			if tc.wantErr {
				if !errors.Is(err, ErrPeerUnidentified) {
					t.Fatalf("err = %v, want ErrPeerUnidentified", err)
				}
				if got.Kind != PeerUnknown {
					t.Fatalf("a refused peer must stay unknown, got %v", got.Kind)
				}
				return
			}
			if err != nil {
				t.Fatalf("unexpected error: %v", err)
			}
			if got.Kind != tc.wantKind {
				t.Fatalf("kind = %v, want %v", got.Kind, tc.wantKind)
			}
			if got.ServiceID != tc.wantID {
				t.Fatalf("service id = %q, want %q", got.ServiceID, tc.wantID)
			}
		})
	}
}

func TestIdentifyPeerPrefersTheSPIFFEURIAmongSeveral(t *testing.T) {
	t.Parallel()

	ca := newInboundCA(t)
	other, err := url.Parse("https://example.test/ignored")
	if err != nil {
		t.Fatalf("parse URI: %v", err)
	}
	spiffe, err := url.Parse(connection.FormatSPIFFE(connection.Identity{ServiceID: "svc-a", InstanceID: "inst-1"}))
	if err != nil {
		t.Fatalf("parse SPIFFE URI: %v", err)
	}
	leaf := ca.issue(t, &x509.Certificate{
		Subject: pkix.Name{CommonName: "servicebridge-leaf"},
		URIs:    []*url.URL{other, spiffe},
	})

	got, err := IdentifyPeer(leaf.Leaf)
	if err != nil {
		t.Fatalf("IdentifyPeer: %v", err)
	}
	if got.Kind != PeerService || got.ServiceID != "svc-a" || got.InstanceID != "inst-1" {
		t.Fatalf("got %+v", got)
	}
}

func TestPeerFromContextWithoutTLS(t *testing.T) {
	t.Parallel()

	if _, err := PeerFromContext(context.Background()); !errors.Is(err, ErrPeerUnidentified) {
		t.Fatalf("err = %v, want ErrPeerUnidentified", err)
	}
}

func TestAllow(t *testing.T) {
	t.Parallel()

	service := Peer{Kind: PeerService, ServiceID: "svc-a", InstanceID: "inst-1"}
	runtimeProxy := Peer{Kind: PeerRuntime}

	rule := func(action, peerID, target string) *pb.PolicyRule {
		return &pb.PolicyRule{Action: action, PeerServiceId: peerID, TargetName: target}
	}

	tests := []struct {
		name    string
		peer    Peer
		method  string
		policy  *pb.PolicyEvaluation
		wantErr error
	}{
		{
			name:   "no evaluation yet allows",
			peer:   service,
			method: "charge",
		},
		{
			name:   "empty rule set allows",
			peer:   service,
			method: "charge",
			policy: &pb.PolicyEvaluation{},
		},
		{
			name:   "rules for other actions do not gate rpc",
			peer:   service,
			method: "charge",
			policy: &pb.PolicyEvaluation{Acceptance: []*pb.PolicyRule{rule("event.handle", "svc-z", "orders.*")}},
		},
		{
			name:   "exact peer and method",
			peer:   service,
			method: "charge",
			policy: &pb.PolicyEvaluation{Acceptance: []*pb.PolicyRule{rule(actionRPCHandle, "svc-a", "charge")}},
		},
		{
			name:   "wildcard peer",
			peer:   service,
			method: "charge",
			policy: &pb.PolicyEvaluation{Acceptance: []*pb.PolicyRule{rule(actionRPCHandle, "", "charge")}},
		},
		{
			name:   "wildcard method",
			peer:   service,
			method: "charge",
			policy: &pb.PolicyEvaluation{Acceptance: []*pb.PolicyRule{rule(actionRPCHandle, "svc-a", wildcardTarget)}},
		},
		{
			name:    "another peer is denied",
			peer:    service,
			method:  "charge",
			policy:  &pb.PolicyEvaluation{Acceptance: []*pb.PolicyRule{rule(actionRPCHandle, "svc-b", wildcardTarget)}},
			wantErr: ErrAcceptanceDenied,
		},
		{
			name:    "another method is denied",
			peer:    service,
			method:  "refund",
			policy:  &pb.PolicyEvaluation{Acceptance: []*pb.PolicyRule{rule(actionRPCHandle, "svc-a", "charge")}},
			wantErr: ErrAcceptanceDenied,
		},
		{
			name:   "runtime proxy passes a policy that denies every service",
			peer:   runtimeProxy,
			method: "refund",
			policy: &pb.PolicyEvaluation{Acceptance: []*pb.PolicyRule{rule(actionRPCHandle, "svc-b", "charge")}},
		},
		{
			name:    "unidentified peer is refused",
			peer:    Peer{},
			method:  "charge",
			wantErr: ErrPeerUnidentified,
		},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			t.Parallel()

			err := Allow(tc.peer, tc.method, tc.policy)
			switch {
			case tc.wantErr == nil && err != nil:
				t.Fatalf("unexpected denial: %v", err)
			case tc.wantErr != nil && !errors.Is(err, tc.wantErr):
				t.Fatalf("err = %v, want %v", err, tc.wantErr)
			}
		})
	}
}
